### DISCUSSION TITLE
Possible to install a specific version? Or how to now, which version I'll get?

### DISCUSSION THREAD

**Discussion by mfn** [shivammathur/setup-php#453] [Q&A / Help] [answered]
Sorry, stupid question probably.

So currently there's a super breaking change going on with 7.4.18 and 8.0.5:
- https://bugs.php.net/bug.php?id=81002
- https://github.com/php/php-src/pull/6801#issuecomment-828783810

(things are already in progress of being reverted, so with the release after 7.4.18/8.0.5 we'll be good)

Now, I thought I was (too?) clever and tried `php-version: 7.4.18` to test the extend on my codebase, because I noticed my setup with `php-version: 7.4` still gives me 7.4.16, alas it doesn't work that way 😅 I've been using setup-php for too long to realize only the major/minor at are supported, as documented at https://github.com/shivammathur/setup-php#tada-php-support

Still I'm now curious:
- how could I know in _advance_ what version I get with e.g. `7.4`, specifically? (same for 8.0)
- is it planned to support arbitrary compilation of specific version?

Thanks for all your hard work!

**Comment by shivammathur** [comment, marked as answer]
For PHP 7.4 and 8.0, you will get the versions which are pre-installed on the images i.e. `7.4.16` and `8.0.3`. Once the runner images are updated you will get the latest patch versions `7.4.18` and `8.0.5`.

If you specify the environment variable `update: true` then setup-php will update the php on the images to `7.4.18` and `8.0.5` respectively. This update process now uses cached builds of `ppa:ondrej/php`, so should only take a second or two more.

Test without `update: true` : https://github.com/shivammathur/test-setup-php/actions/runs/806975891/workflow
Test with `update: true` : https://github.com/shivammathur/test-setup-php/actions/runs/806979414/workflow

**Reply by shivammathur** [reply]
Supporting the exact semver would mean maintaining builds of each patch release for all the supported versions, which would be a huge task and out of scope for this project.

**Reply by shivammathur** [reply]
I have rebuilt the cache after ondrej added the revert patch for the pgsql bug.

**Reply by mfn** [reply]
Thank you!

**Reply by danon** [reply]
@shivammathur What should I do when I have PHP-version specific code, which works differently between PHP patch versions and I'd like to run CI for both of then?

**Reply by shivammathur** [reply]
@Danon 
Technically, you can use the [php docker images](https://hub.docker.com/_/php) as containers and create a workflow that tests on specific patch versions.
But, if your code works differently between patch versions, that would be a breaking change in PHP code. I would recommend you report the breaking change to https://bugs.php.net/

setup-php as I have said earlier cannot support specifying exact PHP patch versions as that will increase the maintenance work with each patch release.

**Reply by danon** [reply]
@shivammathur Yes, that's exactly what happens. PHP introduces breaking changes between patch versions. I've worked with those breaking changes in my application, doing all the `if`-ology, to minimise the differences, and I have a set of unit tests working on different php versions. I had no problem with running that on Travis.

> setup-php as I have said earlier cannot support specifying exact PHP patch versions as that will increase the maintenance work with each patch release.

You don't have to add *each* new patch version. Maybe the lowest and highest would do? Like Only `7.3.0` and `7.3.36`, but nothing in between?

**Reply by mfn** [reply]
>  I had no problem with running that on Travis.

How did they manage it? Did they roll their own management for this or where they using external (re)sources for this?

**Reply by danon** [reply]
I think they used PhpEnv project (https://github.com/phpenv/phpenv), and downloaded binaries to agents

For example for PHP 8.0.1 they downloaded (probably using `curl`):
 - https://storage.googleapis.com/travis-ci-language-archives/php/binaries/ubuntu/16.04/x86_64/php-8.0.1.tar.bz2 and stored it in `/home/travis/.phpenv/versions/8.0.1`
 - https://storage.googleapis.com/travis-ci-language-archives/php/binaries/ubuntu/16.04/x86_64/php-8.0.2.tar.bz2 and stored it in `/home/travis/.phpenv/versions/8.0.2`

Then they called `phpenv` to set proper version
```
$ phpenv global 8.0.2
```
and then you can use `php@8.0.2`
```
$ php --version
PHP 8.0.2 (cli) (built: Feb  7 2021 11:49:25) ( ZTS )
```

**Reply by shivammathur** [reply]
@Danon 

This is not a technical issue, but a maintenance and security issue.

Yes, TravisCI provides patch versions, but they are a company with resources. This work is open-source and currently does not have enough resources to maintain these.

If I add support for one patch version, I can add support for all patch versions as once there is setup for one, it can build all. The issue would be maintaining this and backporting security patches to older patch versions of PHP. This in currently out of scope of this project.

In future, I would like to add support for patch versions, zts, debug, Asan and UBsan builds. These are all milestones for this project and would need substantially more funding.
